services/dedup/dedup.go

```diff
@@ -0,0 +1,233 @@
+			_, err := txn.Get([]byte(messageID))
+			if err != badger.ErrKeyNotFound {
+				toRemoveMessageIndexesSet[idx] = struct{}{}
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		panic(err)
+	}
+
+	toRemoveMessageIndexes := make([]int, 0, len(toRemoveMessageIndexesSet))
+	for k := range toRemoveMessageIndexesSet {
+		toRemoveMessageIndexes = append(toRemoveMessageIndexes, k)
+	}
+
+	sort.Ints(toRemoveMessageIndexes)
+	count := 0
+	for _, idx := range toRemoveMessageIndexes {
+		pkgLogger.Debugf("Dropping event with duplicate messageId: %s", messageIDs[idx])
+		misc.IncrementMapByKey(sourceDupStats, writeKey, 1)
+		*body, err = sjson.DeleteBytes(*body, fmt.Sprintf(`batch.%v`, idx-count))
+		if err != nil {
+			panic(err)
+		}
+		count++
+	}
+}
+
+func (d *DedupHandleT) updateSourceStats(sourceStats map[string]int, bucket string) {
+	for sourceTag, count := range sourceStats {
+		tags := map[string]string{
+			"source": sourceTag,
+		}
+		sourceStatsD := d.stats.NewTaggedStat(bucket, stats.CountType, tags)
+		sourceStatsD.Count(count)
+	}
+}
+
+func addToSet(set map[string]struct{}, elements []string) {
+	for _, element := range elements {
+		set[element] = struct{}{}
+	}
+}
+
+//Takes a job list, dedupes them in place and finally returns the modified list
+//duplicated jobs are moved to succeded state, so that they are not picked in next turn
+func (d *DedupHandleT) DedupJobs(jobList []*jobsdb.JobT) []*jobsdb.JobT {
+	if len(jobList) > 0 {
+		var dedupedJobList []*jobsdb.JobT
+		var duplicateList []*jobsdb.JobT
+		var sourceDupStats = make(map[string]int)
+
+		allMessageIdsSet := make(map[string]struct{})
+
+		for _, unprocessedJob := range jobList {
+			body, err := unprocessedJob.EventPayload.MarshalJSON()
+			if err != nil {
+				pkgLogger.Infof("Marshalling job payload failed with : %s", err.Error())
+				continue
+			}
+			result := gjson.GetBytes(body, "batch")
+			var index int
+			var reqMessageIDs []string
+			result.ForEach(func(_, _ gjson.Result) bool {
+				reqMessageIDs = append(reqMessageIDs, gjson.GetBytes(body, fmt.Sprintf(`batch.%v.messageId`, index)).String())
+				index++
+				return true // keep iterating
+			})
+
+			writeKey := gjson.GetBytes(body, "writeKey").Str
+			d.dedupMessages(&body, reqMessageIDs, writeKey, sourceDupStats)
+			if len(gjson.GetBytes(body, "batch").Array()) == 0 {
+				duplicateList = append(duplicateList, unprocessedJob)
+				continue
+			}
+			unprocessedJob.EventPayload.UnmarshalJSON(body)
+			addToSet(allMessageIdsSet, reqMessageIDs)
+			dedupedJobList = append(dedupedJobList, unprocessedJob)
+		}
+		if len(dedupedJobList) > 0 {
+			messageIdsArr := make([]string, 0)
+			for msgId := range allMessageIdsSet {
+				messageIdsArr = append(messageIdsArr, msgId)
+			}
+			rruntime.Go(func() {
+				d.writeToBadger(messageIdsArr)
+			})
+		}
+		//Mark all as executing so next query doesn't pick it up
+		var statusList []*jobsdb.JobStatusT
+		for _, batchEvent := range duplicateList {
```

Comment:
Dedup should not be directly writing to jobsDB. Instead, it should only work on top of badger.

Reply:
It would be ideal if we could restrict jobsdb transactions only to the processor.